core/src/main/java/hudson/util/DescribableList.java

```diff
@@ -270,10 +270,9 @@ public void marshal(Object source, HierarchicalStreamWriter writer, MarshallingC
         }
 
         public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
-            CopyOnWriteList core = copyOnWriteListConverter.unmarshal(reader, context);
-
             try {
-                DescribableList r = (DescribableList)context.getRequiredType().newInstance();
+                DescribableList r = (DescribableList) context.getRequiredType().asSubclass(DescribableList.class).newInstance();
+                CopyOnWriteList core = copyOnWriteListConverter.unmarshal(reader, context);
```

Comment:
The workaround consists of moving the `unmarshal` call _after_ the `getRequiredType` call, so even if `unmarshal` corrupts this aspect of the `context`, we can still produce a meaningful error.